kv/etcd/etcd.go

```diff
@@ -0,0 +1,292 @@
+// THE SOFTWARE.
+
+package etcd
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/golang/protobuf/proto"
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/retry"
+	"github.com/uber-go/tally"
+	"golang.org/x/net/context"
+)
+
+var noopCancel func()
+
+// NewStore creates a kv store based on etcd
+func NewStore(etcd *clientv3.Client, opts Options) kv.Store {
+	scope := opts.InstrumentsOptions().MetricsScope()
+	return &client{
+		opts:       opts,
+		kv:         etcd.KV,
+		watcher:    etcd.Watcher,
+		watchables: map[string]kv.ValueWatchable{},
+		retrier:    xretry.NewRetrier(opts.RetryOptions()),
+		logger:     opts.InstrumentsOptions().Logger(),
+		m: clientMetrics{
+			etcdGetError:    scope.Counter("etcd-get-error"),
+			etcdPutError:    scope.Counter("etcd-put-error"),
+			etcdTnxError:    scope.Counter("etcd-tnx-error"),
+			etcdWatchCreate: scope.Counter("etcd-watch-create"),
+			etcdWatchError:  scope.Counter("etcd-watch-error"),
+			etcdWatchReset:  scope.Counter("etcd-watch-reset"),
+		},
+	}
+}
+
+type client struct {
+	sync.RWMutex
+
+	opts       Options
+	kv         clientv3.KV
+	watcher    clientv3.Watcher
+	watchables map[string]kv.ValueWatchable
+	retrier    xretry.Retrier
+	logger     xlog.Logger
+	m          clientMetrics
+}
+
+type clientMetrics struct {
+	etcdGetError    tally.Counter
+	etcdPutError    tally.Counter
+	etcdTnxError    tally.Counter
+	etcdWatchCreate tally.Counter
+	etcdWatchError  tally.Counter
+	etcdWatchReset  tally.Counter
+}
+
+func (c *client) Get(key string) (kv.Value, error) {
+	ctx, cancel := c.context()
+	defer cancel()
+
+	r, err := c.kv.Get(ctx, c.opts.KeyFn()(key))
+	if err != nil {
+		c.m.etcdGetError.Inc(1)
+		return nil, err
+	}
+
+	if r.Count == 0 {
+		return nil, kv.ErrNotFound
+	}
+
+	if r.Count > 1 {
+		return nil, fmt.Errorf("received %d values for key %s, expecting 1", r.Count, key)
+	}
+
+	return kv.NewValue(r.Kvs[0].Value, int(r.Kvs[0].Version)), nil
+}
+
+func (c *client) Watch(key string) (kv.ValueWatch, error) {
+	c.Lock()
+	watchable, ok := c.watchables[key]
+	if !ok {
+		watchChan := c.watcher.Watch(
+			context.Background(),
+			c.opts.KeyFn()(key),
+			clientv3.WithProgressNotify(),
+			// Receive initial notification once the watch channel is created
+			clientv3.WithCreatedNotify(),
+		)
+		c.m.etcdWatchCreate.Inc(1)
+
+		watchable = kv.NewValueWatchable()
+		c.watchables[key] = watchable
+
+		go func() {
+			ticker := time.Tick(c.opts.WatchChanCheckInterval())
+
+			for {
+				select {
+				case r, ok := <-watchChan:
+					if ok {
+						c.processNotification(r, watchable, key)
+					} else {
+						c.logger.Warnf("etcd watch channel closed on key %s, recreating a watch channel", key)
+
+						// avoid recreating watch channel too frequently
+						time.Sleep(c.opts.WatchChanResetInterval())
+
+						watchChan = c.watcher.Watch(
```

Comment:
> Is there any option on the Watch call to not get those information, only get notifications? If not, probably worth adding one?

No. This is not the exact use case we designed the watch API for (the current API is richer in feature). If you worry about the additional cost by sending the events, can you create an issue in etcd repo, explaining your use case and letting the etcd team know why you need the option?

Thanks.